File: python/canopy_ai/resolve.py
from urllib.parse import quote

from canopy_ai.errors import CanopyApiError
from canopy_ai.transport import Transport


def resolve_entity(transport: Transport, slug: str) -> str:
    """Resolves a registry slug (e.g. agentic.market/anthropic) to an address."""
    _, body = transport.request(
        "GET",
        f"/api/resolve?entity={quote(slug, safe='')}",
        expect_statuses=[200],
    )
    if not isinstance(body, dict) or not body.get("address"):
        raise CanopyApiError(200, f'Entity "{slug}" has no resolved address', body)
    return str(body["address"])

File: typescript/src/resolve.ts
import type { Transport } from "./transport.js";
import { CanopyApiError } from "./errors.js";

interface ResolveResponse {
  address?: string;
  entity?: unknown;
}

/**
 * Resolves an entity-registry slug (e.g. `agentic.market/anthropic`) to an
 * on-chain address via `/api/resolve`. Throws if the entity isn't found.
 */
export async function resolveEntity(transport: Transport, slug: string): Promise<string> {
  const { body } = await transport.request<ResolveResponse>({
    method: "GET",
    path: `/api/resolve?entity=${encodeURIComponent(slug)}`,
    expectStatuses: [200],
  });
  if (!body.address) {
    throw new CanopyApiError(200, `Entity "${slug}" has no resolved address`, body);
  }
  return body.address;
}
